<a href="https://colab.research.google.com/github/siddiqial/StartupCampus/blob/main/SC62110_Siddiqi_Albiruny_Data_Preprocessing_(Sales_Marketing).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
data = pd.read_csv("/content/train.csv")
display(data)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

- SalePrice - harga jual properti dalam dolar. Ini adalah variabel target yang ingin Anda prediksi.
- MSSubClass: Kelas bangunan
- MSZoning: Klasifikasi zonasi umum
- LotFrontage: Kaki jalan linier yang terhubung ke properti
- LotArea: Ukuran lot dalam kaki persegi
- Street: Jenis akses jalan
- Alley: Jenis akses gang
- LotShape: Bentuk umum properti
- LandContour: Kerataan properti
- Utilites: Jenis utilitas yang tersedia
- LotConfig: Konfigurasi banyak
- LandSlope: Kemiringan properti
- Neighborhood: Lokasi fisik dalam batas kota Ames
- Condition 1: Dekat dengan jalan utama atau rel kereta api
- Condition 2: Dekat dengan jalan utama atau rel kereta api (jika ada yang kedua)
- BldgType: Tipe hunian
- HouseStyle: Gaya hunian
- OverallQual: Keseluruhan material dan kualitas hasil akhir
- OverallCond: Peringkat kondisi keseluruhan
- YearBuilt: Tanggal konstruksi asli
- YearRemodAdd: Tanggal perombakan
- RoofStyle: Jenis atap
- RoofMatl : Bahan atap
- Exterior1st: Penutup eksterior pada rumah
- Exterior2nd: Penutup eksterior pada rumah (jika lebih dari satu bahan)
- MasVnrType: Jenis veneer pasangan bata
- MasVnrArea: Area veneer batu dalam kaki persegi
- ExterQual: Kualitas material eksterior
- ExterCond: Kondisi material eksterior saat ini
- Pondasi: Jenis pondasi
- BsmtQual : Ketinggian basement
- BsmtCond : Kondisi umum basement
- BsmtExposure: Dinding basement walkout atau tingkat taman
- BsmtFinType1: Kualitas area basement jadi
- BsmtFinSF1: Tipe 1 kaki persegi jadi
- BsmtFinType2: Kualitas area jadi kedua (jika ada)
- BsmtFinSF2: Tipe 2 kaki persegi jadi
- BsmtUnfSF: Area basement seluas satu kaki persegi yang belum selesai
- TotalBsmtSF: Total kaki persegi luas basement
- Heating: Jenis pemanasan
- HeatingQC: Kualitas dan kondisi pemanasan
- CentralAir: AC sentral
- Elictrical: Sistem kelistrikan
- 1stFlrSF: Kaki persegi Lantai Pertama
- 2ndFlrSF: Lantai dua kaki persegi
- LowQualFinSF: Kaki persegi jadi berkualitas rendah (semua lantai)
- GrLivArea: Kaki persegi ruang tamu di atas permukaan tanah (tanah).
- BsmtFullBath: Kamar mandi lengkap di ruang bawah tanah
- BsmtHalfBath: Setengah kamar mandi bawah tanah
- FullBath: Kamar mandi lengkap di atas kelas
- HalfBath: Setengah kamar mandi di atas kelas
- Bedroom: Jumlah kamar tidur di atas lantai basement
- Kitchen: Jumlah dapur
- KitchenQual : Kualitas dapur
- TotRmsAbvGrd : Total kamar diatas kelas (tidak termasuk kamar mandi)
- Functional: Peringkat fungsionalitas rumah
- Fireplaces: Jumlah perapian
- FireplaceQu: Kualitas perapian
- GarageType: Lokasi garasi
- GarageYrBlt: Tahun garasi dibangun
- GarageFinish: Finishing interior garasi
- GarageCars: Ukuran garasi dalam kapasitas mobil
- GarageArea: Ukuran garasi dalam kaki persegi
 GarageQual: Kualitas garasi
- GarageCond: Kondisi garasi
- PavedDrive: Jalan masuk beraspal
- WoodDeckSF: Luas dek kayu dalam kaki persegi
- OpenPorchSF: Area teras terbuka dalam satuan kaki persegi
- EnclosedPorch: Area teras tertutup dalam ukuran kaki persegi
- 3SsnPorch: Area teras tiga musim dalam kaki persegi
- ScreenPorch: Area teras layar dalam kaki persegi
- PoolArea: Area kolam dalam kaki persegi
- PoolQC: Kualitas kolam
- Fence: Kualitas pagar
- MiscFeature: Fitur lain-lain yang tidak tercakup dalam kategori lain
- MiscVal: $Nilai fitur lain-lain
- MoSold: Bulan Terjual
- YrSold: Tahun Terjual
- SaleType: Jenis penjualan
- SaleCondition: Kondisi penjualan

## Missing Value

In [ ]:
data.info()

Total data yang seharusnya ada adalah 1460. Dalam data ini terdapat beberapa kolom yang datanya Missing. Diantaranya adalah :
- LotFrontage : 259 Data Missing Value
- Alley : 1369 Data Missing Value
- MasVnrType : 8 Data Missing Value
- MasVnrArea : 8 Data Missing Value
- BsmtQual : 37 Data Missing Value
- BsmtCond : 37 Data Missing Value
- BsmtExposure : 38 Data Missing Value
- BsmtFinType1 : 37 Data Missing Value
- BsmtFinType2 : 38 Data Missing Value
- Electrical : 1 Data Missing Value
- FireplaceQu : 690 Data Missing Value
- GarageType : 81 Data Missing Value
- GarageYrBlt : 81 Data Missing Value
- GarageFinish : 81 Data Missing Value
- GarageQual : 81 Data Missing Value
- GarageCond : 81 Data Missing Value
- PoolQC : 1453 Data Missing Value
- Fence : 1179 Data Missing Value
- MiscFeature : 1406 Data Missing Value

Dari sini saya membuat ketentuan yaitu jika data Missing Value mencapai lebih dari 1/4 jumlah data, maka saya akan menghapusnya. Karena sudah tidak merepresentasikan data. Jika datanya <= 1/4 jumlah data maka saya akan mengisinya menggunakan mean atau median.

In [ ]:
data["OverallCond"]

0       5
1       8
2       5
3       5
4       5
       ..
1455    5
1456    6
1457    9
1458    6
1459    6
Name: OverallCond, Length: 1460, dtype: int64

In [19]:
data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  \
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000   
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890   
std      40.177307    496.123024     2.703626     1.328095   79442.502883   
min       0.000000      0.000000     1.000000  2006.000000   34900.000000   
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000   
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000   
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000   
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000   

       TotalLuasBangunan  UsiaBangunan  JumlahKamarMandi  TotalLuasTeras  \
count        1460.000000   1460.000000       1460.000000     1460.000000   
mean         2567.048630     36.547945          1.947945      181.329452   
std           821.714421     30.250152          0.794965      156.656097   
min           334.000000      0.000000          0.000000        0.000000   
25%          2009.500000      8.000000          1.000000       45.000000   
50%          2474.000000     35.000000          2.000000      164.000000   
75%          3004.000000     54.000000          3.000000      266.000000   
max         11752.000000    136.000000          4.000000     1027.000000   

       RatingKualitasRumah  
count          1460.000000  
mean              5.837329  
std               0.846764  
min               1.000000  
25%               5.500000  
50%               6.000000  
75%               6.500000  
max               9.500000  

[8 rows x 43 columns]

In [ ]:
data.describe(include=object)

In [ ]:
meanLotFrontage = np.mean(data.LotFrontage)
print("LotFrontage :", meanLotFrontage)

Disini saya mengatasi Missing Value pada kolom LotFrontage yaitu dengan mengisi nya menggunakan median

In [ ]:
data["LotFrontage"] = data["LotFrontage"].fillna(data["LotFrontage"].median())
data["LotFrontage"]

Disini saya menghapus kolom Alley karena data yang Missing Value lebih dari 1/4 jumlah data

In [ ]:
data = data.drop('Alley', axis=1)

In [ ]:
display(data)

In [ ]:
kolom_MasVnrType = data['MasVnrType']
print(kolom_MasVnrType)

In [ ]:
import statistics as stats
stats.mode(data['MasVnrType'])

In [ ]:
mode_MasVnrType = data['MasVnrType'].mode()[0]  # Mendapatkan mode (nilai paling sering muncul)
data['MasVnrType'] = data['MasVnrType'].fillna(mode_MasVnrType)

Disini saya menggunakan Mode untuk mengisi data Missing Value pada kolom MasVnrType

In [ ]:
meanMasVnrArea = np.mean(data.MasVnrArea)
print("MasVnrArea :", meanMasVnrArea)

In [ ]:
mode_MasVnrArea = data['MasVnrArea'].mode()[0]  # Mendapatkan mode (nilai paling sering muncul)
data['MasVnrArea'] = data['MasVnrArea'].fillna(mode_MasVnrArea)

Disini saya menggunakan mode untuk mengisi Missing Value

In [ ]:
kolom_BsmtQual = data['BsmtQual']
print(kolom_BsmtQual)

In [ ]:
display(data)

In [ ]:
stats.mode(data['BsmtQual'])

In [ ]:
mode_BsmtQual = data['BsmtQual'].mode()[0]  # Mendapatkan mode (nilai paling sering muncul)
data['BsmtQual'] = data['BsmtQual'].fillna(mode_BsmtQual)

In [ ]:
data['BsmtCond']

In [ ]:
mode_BsmtCond = data['BsmtCond'].mode()[0]
data['BsmtCond'] = data['BsmtCond'].fillna(mode_BsmtCond)

In [ ]:
data['BsmtExposure']

NameError: name 'data' is not defined

In [ ]:
stats.mode(data['BsmtExposure'])

In [ ]:
mode_BsmtExposure = data['BsmtExposure'].mode()[0]
data['BsmtExposure'] = data['BsmtExposure'].fillna(mode_BsmtExposure)

In [ ]:
data['BsmtFinType1']

In [ ]:
stats.mode(data['BsmtFinType1'])

In [ ]:
mode_BsmtFinType1 = data['BsmtFinType1'].mode()[0]
data['BsmtFinType1'] = data['BsmtFinType1'].fillna(mode_BsmtFinType1)

In [ ]:
stats.mode(data['BsmtFinType2'])

In [ ]:
mode_BsmtFinType2 = data['BsmtFinType2'].mode()[0]
data['BsmtFinType2'] = data['BsmtFinType2'].fillna(mode_BsmtFinType2)

In [ ]:
data['Electrical']

In [ ]:
stats.mode(data['Electrical'])

In [ ]:
mode_Electrical = data['Electrical'].mode()[0]
data['Electrical'] = data['Electrical'].fillna(mode_Electrical)

In [ ]:
data.info()

In [ ]:
FireplaceQu_TA = data.loc[data['FireplaceQu'] == 'TA', ['FireplaceQu']]
FireplaceQu_TA

In [ ]:
FireplaceQu_Gd = data.loc[data['FireplaceQu'] == 'Gd', ['FireplaceQu']]
FireplaceQu_Gd

In [ ]:
FireplaceQu_Ex = data.loc[data['FireplaceQu'] == 'Ex', ['FireplaceQu']]
FireplaceQu_Ex

In [ ]:
stats.mode(data['FireplaceQu'])

## Feature

- Total luas bangunan : 1stFlrSF + 2ndFlrSF + TotalBsmtSF.
- Usia bangunan : YrSold + YrBuilt
- Jumlah kamar mandi : FullBath + HalfBath
- Total luas teras : WoodDeckSF + OpenPorchSF + EnclosedPorch + 3SsnPorch + ScreenPorch.
- Rating Kualitas rumah : (OverallQual + OverallCond)/2
- Total ruangan : Bedroom + TotRmsAbvGrd
- Kondisi Garasi : (GarageQual + GarageCond) /2
-


In [9]:
# Total Luas Bangunan

data["TotalLuasBangunan"] = data["1stFlrSF"] + data["2ndFlrSF"] + data["TotalBsmtSF"]
data["TotalLuasBangunan"]

0       2566
1       2524
2       2706
3       2473
4       3343
        ... 
1455    2600
1456    3615
1457    3492
1458    2156
1459    2512
Name: TotalLuasBangunan, Length: 1460, dtype: int64

In [11]:
# Usia Bangunan

data["UsiaBangunan"] = data["YrSold"] - data["YearBuilt"]
data["UsiaBangunan"]

0        5
1       31
2        7
3       91
4        8
        ..
1455     8
1456    32
1457    69
1458    60
1459    43
Name: UsiaBangunan, Length: 1460, dtype: int64

In [13]:
# Jumlah Kamar Mandi

data["JumlahKamarMandi"] = data["FullBath"] + data["HalfBath"]
data["JumlahKamarMandi"]

0       3
1       2
2       3
3       1
4       3
       ..
1455    3
1456    2
1457    2
1458    1
1459    2
Name: JumlahKamarMandi, Length: 1460, dtype: int64

In [14]:
# Total Luas Teras

data["TotalLuasTeras"] = data["WoodDeckSF"] + data["OpenPorchSF"] + data["EnclosedPorch"] + data["3SsnPorch"] + data["ScreenPorch"]
data["TotalLuasTeras"]

0        61
1       298
2        42
3       307
4       276
       ... 
1455     40
1456    349
1457     60
1458    478
1459    804
Name: TotalLuasTeras, Length: 1460, dtype: int64

In [15]:
# Rating Kualitas Rumah

data["RatingKualitasRumah"] = (data["OverallQual"] + data["OverallCond"])/2
data["RatingKualitasRumah"]

0       6.0
1       7.0
2       6.0
3       6.0
4       6.5
       ... 
1455    5.5
1456    6.0
1457    8.0
1458    5.5
1459    5.5
Name: RatingKualitasRumah, Length: 1460, dtype: float64

In [21]:
# Total Ruangan

data["TotalRuangan"] = data["BedroomAbvGr"] + data["TotRmsAbvGrd"]
data["TotalRuangan"]

0       11
1        9
2        9
3       10
4       13
        ..
1455    10
1456    10
1457    13
1458     7
1459     9
Name: TotalRuangan, Length: 1460, dtype: int64

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 87 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   1460 non-null   int64  
 1   MSSubClass           1460 non-null   int64  
 2   MSZoning             1460 non-null   object 
 3   LotFrontage          1201 non-null   float64
 4   LotArea              1460 non-null   int64  
 5   Street               1460 non-null   object 
 6   Alley                91 non-null     object 
 7   LotShape             1460 non-null   object 
 8   LandContour          1460 non-null   object 
 9   Utilities            1460 non-null   object 
 10  LotConfig            1460 non-null   object 
 11  LandSlope            1460 non-null   object 
 12  Neighborhood         1460 non-null   object 
 13  Condition1           1460 non-null   object 
 14  Condition2           1460 non-null   object 
 15  BldgType             1460 non-null   o

In [25]:
data[["GarageQual", "GarageCond"]]

GarageQual GarageCond
0            TA         TA
1            TA         TA
2            TA         TA
3            TA         TA
4            TA         TA
...         ...        ...
1455         TA         TA
1456         TA         TA
1457         TA         TA
1458         TA         TA
1459         TA         TA

[1460 rows x 2 columns]

In [22]:
# Kondisi Garasi

data["KondisiGarasi"] = (data["GarageQual"] + data["GarageCond"])/2
data["KondisiGarasi"]

TypeError: unsupported operand type(s) for /: 'str' and 'int'